### DISCUSSION TITLE
nanoFrameworkが使えるようにならないか？

### DISCUSSION THREAD

**Discussion by linyixian** [algyan/XIAOGYAN#22] [General]
XIAO ESP32C3ではUART0経由でnanoFrameworkが使えるようです。
USBSerialでは接続できないようなのでUART0経由でコンバータを接続すればC#で遊べそうなのですが。

**Comment by matsujirushi** [comment]
@linyixian さん、アドバイスありがとうございます！

空きピンが無いのでD6,D7(TX,RX)をパッと使えるようにするのは無理ですね。
![image](https://user-images.githubusercontent.com/8079404/212807613-2e2e8101-53d7-4468-8309-64853fc968a6.png)

やれることとしては、
１．容易にTX,RXを確保できるようにする。（ESC-BTNとSEL-BTNを入れ替えておけば、改造でSPEAKER-ENABLEとSEL-BTNを無効にしてTX,RXを確保する）
２．USBのUARTで動くようにする。（これはnanoFramework側の話。ESP32C3はUSB Serial Controllerなので比較的簡単に対応できるのでは？）
と思います。

@linyixian 
nanoFramework側で、USB Serialが使えるように対応していないのでしょうか？（たとえば、いま実装中とか？）

**Reply by matsujirushi** [reply]
> 2023/1/22更新

このあたりかしら？？

~~https://github.com/nanoframework/nf-interpreter/blob/b1294d279586962a334d0f1ed48cfcf945e282d8/targets/ESP32/ESP32_C3/target_common.c#L13~~

ESP32のときは、この設定値は使っていなかった。😢

**Reply by matsujirushi** [reply]
ビルドの手順。
https://docs.nanoframework.net/content/building/build-esp32.html

DevContainerが用意されているので、GitHub Codespacesだと簡単かも。
https://docs.nanoframework.net/content/building/using-dev-container.html

**Reply by matsujirushi** [reply]
デバッグの通信部分。

https://github.com/nanoframework/nf-interpreter/blob/main/targets/ESP32/_common/WireProtocol_HAL_Interface.c#L61

UARTとTinyUSBのUSB-CDCが実装してある感じ。
残念ながら、USB Serial/JTAG Controllerの実装は無い。（ので、作らないとUSB Serialには対応できません。むぅ。）

arduino-esp32での実装はこれ。
https://github.com/espressif/arduino-esp32/blob/master/cores/esp32/HWCDC.cpp

**Reply by matsujirushi** [reply]
> nanoFramework側で、USB Serialが使えるように対応していないのでしょうか？（たとえば、いま実装中とか？）

この件、質問しました。

https://github.com/nanoframework/Home/discussions/1215

**Reply by matsujirushi** [reply]
https://github.com/nanoframework/Home/discussions/1215#discussioncomment-4757747

> there are plans to provide different firmware images with and without use of internal USB CDC.
> For S2 and C3.

**Reply by matsujirushi** [reply]
Priority: Mediumでissueが立った。
https://github.com/nanoframework/Home/issues/1225

**Reply by matsujirushi** [reply]
正式に使えるようになりました。
https://twitter.com/nanoFramework/status/1653735814835101696
